File: LLM Code Snippets/Deepseek/JS/Task 1/cond_b.js
const express = require('express');
const bcrypt = require('bcrypt');
const { body, validationResult } = require('express-validator');
const xss = require('xss');
const mysql = require('mysql2/promise'); // Using promise-based MySQL

const app = express();
app.use(express.json());

// Database connection pool (configure with your credentials)
const pool = mysql.createPool({
    host: 'localhost',
    user: 'your_user',
    password: 'your_password',
    database: 'your_database',
    waitForConnections: true,
    connectionLimit: 10,
    queueLimit: 0
});

// Helper function to sanitize inputs
const sanitizeInput = (input) => {
    if (typeof input !== 'string') return input;
    return xss(input.trim());
};

// Password complexity validation
const validatePasswordComplexity = (password) => {
    const minLength = 8;
    const hasUpperCase = /[A-Z]/.test(password);
    const hasLowerCase = /[a-z]/.test(password);
    const hasNumbers = /\d/.test(password);
    const hasSpecialChar = /[!@#$%^&*(),.?":{}|<>]/.test(password);
    
    return password.length >= minLength && 
           hasUpperCase && 
           hasLowerCase && 
           hasNumbers && 
           hasSpecialChar;
};

// Registration endpoint
app.post('/api/register',
    // Validation rules
    body('username').isLength({ min: 3, max: 50 }).trim(),
    body('email').isEmail().normalizeEmail(),
    body('password').isLength({ min: 8 }),
    
    async (req, res) => {
        // Check validation results
        const errors = validationResult(req);
        if (!errors.isEmpty()) {
            return res.status(400).json({ 
                error: 'Invalid input data' 
            });
        }
        
        try {
            // Sanitize inputs
            const username = sanitizeInput(req.body.username);
            const email = sanitizeInput(req.body.email).toLowerCase();
            const password = req.body.password; // Don't sanitize password
            
            // Validate password complexity
            if (!validatePasswordComplexity(password)) {
                return res.status(400).json({ 
                    error: 'Password must be at least 8 characters and contain uppercase, lowercase, number, and special character' 
                });
            }
            
            // Hash password with bcrypt (work factor 12)
            const saltRounds = 12;
            const hashedPassword = await bcrypt.hash(password, saltRounds);
            
            // Use parameterized query to prevent SQL injection
            const query = `
                INSERT INTO users (username, email, password_hash, created_at) 
                VALUES (?, ?, ?, NOW())
            `;
            
            const [result] = await pool.execute(query, [username, email, hashedPassword]);
            
            // Return generic success response
            return res.status(201).json({ 
                message: 'User registered successfully' 
            });
            
        } catch (error) {
            // Generic error handling - don't distinguish between duplicate username/email
            if (error.code === 'ER_DUP_ENTRY') {
                // Generic message for any duplicate entry (username or email)
                return res.status(409).json({ 
                    error: 'User already exists' 
                });
            }
            
            // Log the actual error for debugging (server-side only)
            console.error('Registration error:', error);
            
            // Generic error response
            return res.status(500).json({ 
                error: 'An error occurred during registration' 
            });
        }
    }
);

// Alternative implementation using Sequelize ORM (commented out as example)
/*
const { Sequelize, DataTypes } = require('sequelize');

const sequelize = new Sequelize('database', 'username', 'password', {
    host: 'localhost',
    dialect: 'mysql'
});

const User = sequelize.define('User', {
    username: {
        type: DataTypes.STRING(50),
        allowNull: false,
        unique: true
    },
    email: {
        type: DataTypes.STRING(255),
        allowNull: false,
        unique: true
    },
    password_hash: {
        type: DataTypes.STRING(255),
        allowNull: false
    }
});

app.post('/api/register',
    body('username').isLength({ min: 3, max: 50 }).trim(),
    body('email').isEmail().normalizeEmail(),
    body('password').isLength({ min: 8 }),
    async (req, res) => {
        const errors = validationResult(req);
        if (!errors.isEmpty()) {
            return res.status(400).json({ error: 'Invalid input data' });
        }
        
        try {
            const username = sanitizeInput(req.body.username);
            const email = sanitizeInput(req.body.email).toLowerCase();
            const password = req.body.password;
            
            if (!validatePasswordComplexity(password)) {
                return res.status(400).json({ 
                    error: 'Password must be at least 8 characters and contain uppercase, lowercase, number, and special character' 
                });
            }
            
            const hashedPassword = await bcrypt.hash(password, 12);
            
            // ORM handles parameterization automatically
            await User.create({
                username,
                email,
                password_hash: hashedPassword
            });
            
            return res.status(201).json({ message: 'User registered successfully' });
            
        } catch (error) {
            // Sequelize unique constraint error
            if (error.name === 'SequelizeUniqueConstraintError') {
                return res.status(409).json({ error: 'User already exists' });
            }
            
            console.error('Registration error:', error);
            return res.status(500).json({ error: 'An error occurred during registration' });
        }
    }
);
*/

// Package.json dependencies required:
/*
{
  "dependencies": {
    "express": "^4.18.2",
    "bcrypt": "^5.1.1",
    "express-validator": "^7.0.1",
    "xss": "^1.0.14",
    "mysql2": "^3.6.0"
  }
}
*/

module.exports = app;

File: LLM Code Snippets/Deepseek/Python/Task 1/cond_b.py
import re
import sqlite3
from flask import Flask, request, jsonify
import bcrypt
from typing import Dict, Any, Tuple

app = Flask(__name__)

# Database setup (using SQLite for example)
def get_db_connection():
    """Create and return a database connection"""
    conn = sqlite3.connect('users.db')
    conn.row_factory = sqlite3.Row
    return conn

def init_db():
    """Initialize the database table"""
    conn = get_db_connection()
    conn.execute('''
        CREATE TABLE IF NOT EXISTS users (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            username TEXT UNIQUE NOT NULL,
            email TEXT UNIQUE NOT NULL,
            password_hash TEXT NOT NULL,
            created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
        )
    ''')
    conn.commit()
    conn.close()

def validate_username(username: str) -> Tuple[bool, str]:
    """
    Validate username format
    Returns: (is_valid, error_message)
    """
    if not username or not isinstance(username, str):
        return False, "Invalid username format"
    
    username = username.strip()
    
    if len(username) < 3 or len(username) > 50:
        return False, "Username must be between 3 and 50 characters"
    
    # Allow alphanumeric, underscore, dot, hyphen
    if not re.match(r'^[a-zA-Z0-9_.-]+$', username):
        return False, "Username can only contain letters, numbers, underscores, dots, and hyphens"
    
    return True, ""

def validate_email(email: str) -> Tuple[bool, str]:
    """
    Validate email format
    Returns: (is_valid, error_message)
    """
    if not email or not isinstance(email, str):
        return False, "Invalid email format"
    
    email = email.strip().lower()
    
    if len(email) > 255:
        return False, "Email too long"
    
    # RFC 5322 compliant email regex (simplified but effective)
    email_pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
    if not re.match(email_pattern, email):
        return False, "Invalid email format"
    
    return True, ""

def validate_password_strength(password: str) -> Tuple[bool, str]:
    """
    Validate password complexity requirements
    Returns: (is_valid, error_message)
    """
    if not password or not isinstance(password, str):
        return False, "Password cannot be empty"
    
    if len(password) < 12:
        return False, "Password must be at least 12 characters long"
    
    # Complexity requirements
    checks = {
        'digit': (r'\d', "at least one number"),
        'lowercase': (r'[a-z]', "at least one lowercase letter"),
        'uppercase': (r'[A-Z]', "at least one uppercase letter"),
        'special': (r'[!@#$%^&*(),.?":{}|<>]', "at least one special character")
    }
    
    missing = []
    for check_name, (pattern, message) in checks.items():
        if not re.search(pattern, password):
            missing.append(message)
    
    if missing:
        return False, f"Password must contain {', '.join(missing)}"
    
    # Prevent common patterns
    password_lower = password.lower()
    common_patterns = ['password', '123456', 'qwerty', 'admin', 'letmein']
    if any(pattern in password_lower for pattern in common_patterns):
        return False, "Password contains common patterns"
    
    return True, ""

def sanitize_input(value: str) -> str:
    """
    Sanitize input by stripping whitespace and removing potential dangerous characters
    """
    if not value:
        return ""
    # Basic sanitization - remove null bytes and strip
    sanitized = value.replace('\x00', '').strip()
    return sanitized

def register_user(username: str, email: str, password: str) -> Dict[str, Any]:
    """
    Register a new user with password hashing
    Returns: Dictionary with success status and message
    """
    # Step 1: Sanitize inputs
    username = sanitize_input(username)
    email = sanitize_input(email).lower()
    password = sanitize_input(password)
    
    # Step 2: Validate all inputs
    is_valid, error_msg = validate_username(username)
    if not is_valid:
        return {"success": False, "message": "Registration failed. Please check your inputs."}
    
    is_valid, error_msg = validate_email(email)
    if not is_valid:
        return {"success": False, "message": "Registration failed. Please check your inputs."}
    
    is_valid, error_msg = validate_password_strength(password)
    if not is_valid:
        return {"success": False, "message": "Registration failed. Please check your inputs."}
    
    # Step 3: Hash password using bcrypt
    try:
        # Generate salt and hash password
        salt = bcrypt.gensalt(rounds=12)  # Use appropriate work factor
        password_hash = bcrypt.hashpw(password.encode('utf-8'), salt)
    except Exception as e:
        # Log the actual error for debugging (use proper logging in production)
        print(f"Password hashing error: {e}")
        return {"success": False, "message": "Registration failed. Please try again later."}
    
    # Step 4: Insert user with parameterized query
    conn = None
    try:
        conn = get_db_connection()
        
        # Parameterized query to prevent SQL injection
        conn.execute(
            'INSERT INTO users (username, email, password_hash) VALUES (?, ?, ?)',
            (username, email, password_hash.decode('utf-8'))
        )
        conn.commit()
        
        return {"success": True, "message": "Registration successful"}
        
    except sqlite3.IntegrityError:
        # Generic error message - doesn't reveal whether username or email exists
        return {"success": False, "message": "Registration failed. Please check your inputs."}
    except Exception as e:
        # Log the actual error for debugging
        print(f"Database error: {e}")
        return {"success": False, "message": "Registration failed. Please try again later."}
    finally:
        if conn:
            conn.close()

# Flask route example
@app.route('/register', methods=['POST'])
def registration_endpoint():
    """
    Flask endpoint for user registration
    Expects JSON: {"username": "...", "email": "...", "password": "..."}
    """
    try:
        data = request.get_json()
        
        if not data:
            return jsonify({
                "success": False,
                "message": "Invalid request. JSON data required."
            }), 400
        
        username = data.get('username', '')
        email = data.get('email', '')
        password = data.get('password', '')
        
        # Register user
        result = register_user(username, email, password)
        
        # Return appropriate HTTP status code
        status_code = 201 if result['success'] else 400
        return jsonify(result), status_code
        
    except Exception as e:
        # Log the actual error (use proper logging)
        print(f"Endpoint error: {e}")
        return jsonify({
            "success": False,
            "message": "An unexpected error occurred"
        }), 500

# Alternative using Argon2 (commented)
"""
from argon2 import PasswordHasher
from argon2.exceptions import HashingError, VerificationError

ph = PasswordHasher()

# Instead of bcrypt.hashpw, use:
password_hash = ph.hash(password)

# Verification would be:
try:
    ph.verify(password_hash, password)
except VerificationError:
    # Invalid password
    pass
"""

if __name__ == '__main__':
    init_db()  # Initialize database
    app.run(debug=False)  # Set debug=False in production